### PR TITLE
Add rewrapping a value from one value class to another

### PR DESCRIPTION
**Pull request by ekhov**
Implementation of #249

- [x] Add rewrapping value of one value class to another value class if value is the same type.
      ```case class Foo(t: T) extends AnyVal => case class Bar(t: T) extends AnyVal```

- [x] Add unwrapping of a value class's value of one type(`T`) to another type(`S`) if `T => S` transformer exists.
       ```case class Foo(t: T) extends AnyVal => S```
       
- [x] Add wrapping a value of one type(`T`) to value class of another type(`S`) if `T => S` transformer exists.
       ```T => case class Foo(s: S)```
       
- [x]  Add rewrapping value(type `T`) of one value class to another value class(with value of type `S`  if `T => S` transformer exists.
        ```case class Foo(t: T) => case class Bar(s: S)```

### CODE REVIEW

**Comment by krzemin** [comment]
Thanks @ekhov, looks good so far 👍 

In order to cover the following cases, I guess the easiest way would be to generalize existing value class rules (including the one you have created), so that they work with any recursively derived `S => T` transformation (which should cover also the specific `T => T` case).

`resolveRecursiveTransformerBody` will be your friend.

File: chimney/src/main/scala/io/scalaland/chimney/internal/macros/TransformerMacros.scala [hunk 1]
```diff
@@ -239,21 +241,19 @@ trait TransformerMacros extends MappingMacros with TargetConstructorMacros with
...
+
+    for {
+      fromValueClassMember <- fromValueClassMember
+      fromValueClassMemberType = fromValueClassMember.returnType
+      fromMemberAccessTree = q"$srcPrefixTree.${fromValueClassMember.name}"
+      transformerBodyTree <- resolveRecursiveTransformerBody(fromMemberAccessTree, config)(
+        fromValueClassMemberType,
+        To
+      )
+    } yield mkTransformerBodyTree0(transformerBodyTree.target)(transformerBodyTree.tree)
```
Comment (krzemin) [inline chimney/src/main/scala/io/scalaland/chimney/internal/macros/TransformerMacros.scala:256]: You should use `config.derivationTarget` - this is the most probable reason why the tests are failing.
Same applies to the remaining rules.

Comment (krzemin) [inline chimney/src/main/scala/io/scalaland/chimney/internal/macros/TransformerMacros.scala:256]:
```suggestion
    } yield transformerBodyTree.tree
```
Returning just recursively derived tree should be sufficient in this case.

File: chimney/src/main/scala/io/scalaland/chimney/internal/macros/TransformerMacros.scala [hunk 2]
```diff
@@ -263,11 +263,18 @@ trait TransformerMacros extends MappingMacros with TargetConstructorMacros with
...
+    )
+
+    for {
+      toValueClassMemberType <- toValueClassMember.map(_.returnType)
+      transformerBodyTree <- resolveRecursiveTransformerBody(srcPrefixTree, config)(
+        From,
+        toValueClassMemberType
+      )
+      toCreationTree = q"new $To(${transformerBodyTree.tree})"
+    } yield mkTransformerBodyTree0(transformerBodyTree.target)(toCreationTree)
```
Comment (krzemin) [inline chimney/src/main/scala/io/scalaland/chimney/internal/macros/TransformerMacros.scala:277]:
```suggestion
    for {
      toValueClassMethodSymbol <- toValueClassMember
      toValueClassMemberType <- toValueClassMember.map(_.returnType)
      transformerBodyTree <- resolveRecursiveTransformerBody(srcPrefixTree, config)(
        From,
        toValueClassMemberType
      )
    } yield mkTransformerBodyTree1(
      To,
      Target.fromField(toValueClassMethodSymbol, From),
      transformerBodyTree,
      config.derivationTarget
    )(innerTree => q"new $To(${innerTree})")
```
This one is going to be quite more complicated to get it right.
It's not sufficient to just wrap recursively derived tree in `new To(...)`, as recursive tree might be lifted/partial transformer that you have to map over. For such cases `mkTransformerBodyTree1` should be used as in the snipped.

File: chimney/src/main/scala/io/scalaland/chimney/internal/macros/TransformerMacros.scala [hunk 3]
```diff
@@ -282,6 +289,35 @@ trait TransformerMacros extends MappingMacros with TargetConstructorMacros with
...
+      fromValueClassMember <- fromValueClassMember
+      fromValueClassMemberType = fromValueClassMember.returnType
+      toValueClassMemberType <- toValueClassMember.map(_.returnType)
+      fromMemberAccessTree = q"$srcPrefixTree.${fromValueClassMember.name}"
+      transformerBodyTree <- resolveRecursiveTransformerBody(fromMemberAccessTree, config)(
+        fromValueClassMemberType,
+        toValueClassMemberType
+      )
+      toCreationTree = q"new $To(${transformerBodyTree.tree})"
+    } yield mkTransformerBodyTree0(transformerBodyTree.target)(toCreationTree)
```
Comment (krzemin) [inline chimney/src/main/scala/io/scalaland/chimney/internal/macros/TransformerMacros.scala:318]: This rule may require similar `mkTransformerBodyTree1` adjustment as in `expandTypeToValueClass` rule.
Please add at least at test that checks transformation between two value classes, relying on implicit partial transformer to transform their inner types.

**Review by krzemin — CHANGES_REQUESTED** [review]
Hi, thanks for your submission! I left few comments with suggestions how to fix failing unit tests.

File: chimney/src/main/scala/io/scalaland/chimney/internal/macros/TransformerMacros.scala (hunk 1 above)
Comment (ekhov) [reply chimney/src/main/scala/io/scalaland/chimney/internal/macros/TransformerMacros.scala:256]: @krzemin Thanks for the suggestions! Sorry for not delivering for too long- I managed to figure these ones myself, but didn't have much time to actually make a fix. But I'm just working on the fix right now, so hope today PR will be ready!

**Review by krzemin — APPROVED** [review]
LGTM 👍

**Pull request merged**